cmd/erasure-object.go

```diff
@@ -1385,34 +1385,160 @@ func (er erasureObjects) TransitionObject(ctx context.Context, bucket, object st
 // is restored locally to the bucket on source cluster until the restore expiry date.
 // The copy that was transitioned continues to reside in the transitioned tier.
 func (er erasureObjects) RestoreTransitionedObject(ctx context.Context, bucket, object string, opts ObjectOptions) error {
-	// Acquire a write lock before restoring the object.
-	lk := er.NewNSLock(bucket, object)
-	ctx, err := lk.GetLock(ctx, globalDeleteOperationTimeout)
+	oi, err := er.getObjectInfo(ctx, bucket, object, opts)
 	if err != nil {
 		return err
 	}
-	defer lk.Unlock()
-	oi, err := er.getObjectInfo(ctx, bucket, object, opts)
-	if err != nil {
+
+	if len(oi.Parts) == 1 {
+		var rs *HTTPRangeSpec
+		gr, err := getTransitionedObjectReader(ctx, bucket, object, rs, http.Header{}, oi, opts)
+		if err != nil {
+			return err
+		}
+		defer gr.Close()
+		hashReader, err := hash.NewReader(gr, gr.ObjInfo.Size, "", "", gr.ObjInfo.Size)
+		if err != nil {
+			return err
+		}
+		pReader := NewPutObjReader(hashReader)
+		ropts := putRestoreOpts(bucket, object, opts.Transition.RestoreRequest, oi)
+		ropts.UserDefined[xhttp.AmzRestore] = fmt.Sprintf("ongoing-request=%t, expiry-date=%s", false, opts.Transition.RestoreExpiry.Format(http.TimeFormat))
+		if _, err := er.PutObject(ctx, bucket, object, pReader, ropts); err != nil {
+			return err
+		}
+		return nil
+	}
+	return er.restoreTransitionedObject(ctx, bucket, object, opts)
+}
+
+// update restore status header in the metadata
+func (er erasureObjects) updateRestoreMetadata(ctx context.Context, bucket, object string, objInfo ObjectInfo, opts ObjectOptions, noLock bool, rerr error) error {
+	oi := objInfo.Clone()
+	oi.metadataOnly = true // Perform only metadata updates.
+
+	if rerr == nil {
+		oi.UserDefined[xhttp.AmzRestore] = fmt.Sprintf("ongoing-request=%t, expiry-date=%s", false, opts.Transition.RestoreExpiry.Format(http.TimeFormat))
+	} else { // allow retry in the case of failure to restore
+		oi.UserDefined[xhttp.AmzRestore] = fmt.Sprintf("ongoing-request=%t", false)
+	}
+	if _, err := er.CopyObject(ctx, bucket, object, bucket, object, oi, ObjectOptions{
+		VersionID: oi.VersionID,
+	}, ObjectOptions{
+		VersionID: oi.VersionID,
+		NoLock:    noLock, // lock already taken in restoreTransitionedObject
+	}); err != nil {
+		logger.LogIf(ctx, fmt.Errorf("Unable to update transition restore metadata for %s/%s(%s): %s", bucket, object, oi.VersionID, err))
 		return err
 	}
+	// cascade original error back after setting restore header
+	return rerr
```

Comment:
I think the returning the original error the callers must be done outside of this function. The error return value should be used to indicate failure to update object metadata via CopyObject.